doc/developers-guide/Stanza-routing.md

```diff
@@ -1,76 +1,74 @@
 
-```erlang
-(ejabberd@localhost)2> ets:tab2list(route).
-[{route,<<"vjud.localhost">>,
-        {apply_fun,#Fun<ejabberd_router.2.123745223>}},
- {route,<<"muc.localhost">>,
-        {apply_fun,#Fun<mod_muc.2.63726579>}},
- {route,<<"localhost">>,{apply,ejabberd_local,route}}]
-```
+## 5. `ejabberd_local` to `ejabberd_sm`
 
-Here we see that for a domain "localhost" `ejabberd_local:route()`is called.
-Routing the stanza there means calling `mongoose_local_delivery:do_route/5`, which calls `filter_local_packet` hook and, if passed, runs the fun or applies the handler.
-In most cases, the handler is `ejabberd_local:route/3`.
+`ejabberd_local:process_packet/5` checks if the stanza is addressed to a user or to the server itself. For local users like Bob, `ejabberd_sm:route/4` is called.
 
-## 6. `ejabberd_local` to `ejabberd_sm`
+## 6. `ejabberd_sm`
 
-`ejabberd_local` routes the stanza to `ejabberd_sm` given it has at least a bare JID as the recipient.
+`ejabberd_sm` determines the available resources of the recipient, takes into account their priorities and whether the message is addressed to a particular resource or a bare JID.
+It appropriately replicates (or not) the message and sends it to the recipient's C2S process(es) by calling `mongoose_c2s:route/2`.
+In case no resources are available for delivery (hence no C2S processes to pass the message to), `offline_message_hook` is run.
 
-## 7. `ejabberd_sm`
+As Bob has one online session, the message is sent to the C2S process associated with that session.
 
-`ejabberd_sm` determines the available resources of User B, takes into account their priorities and whether the message is addressed to a particular resource or a bare JID. 
-It appropriately replicates (or not) the message and sends it to the recipient's `ejabberd_c2s` process(es).
+## 7. Recipient's C2S process delivers the message
 
-In case no resources are available for delivery (hence no `ejabberd_c2s` processes to pass the message to), `offline_message_hook` is run.
+The `user_receive_packet` hook is run to notify the rest of the system about the stanza delivery.
+Next, depending on the type of the stanza, one of the following hooks is called:
 
-## 8. `ejabberd_c2s`
+* `user_receive_message` for messages,
+* `user_receive_presence` for presences,
+* `user_receive_iq` for IQ (info/query),
+* `user_receive_xmlel` for other XML elements.
 
-`ejabberd_c2s` verifies the stanza against relevant privacy lists and sends it to the socket.
-`user_receive_packet` hook is run to notify the rest of the system about the stanza delivery to User B.
+Each hook can be handled by multiple modules subscribed to it. These hooks' handlers can stop the routing, e.g. when the stanza is blocked by `mod_privacy`. Finally, the `xmpp_presend_element` hook is called, which can stop the routing as well - otherwise, the stanza is converted to binary data and sent to the recipient's TCP socket.
```

Comment:
I'd maybe add here that `xmpp_presend_element` is where stream management or csi hook in to handle buffering, just as an example 🤔